servers/zts/src/main/java/com/yahoo/athenz/zts/store/RolePrefixTrie.java

```diff
@@ -0,0 +1,122 @@
+ *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
+ *
+ */
+
+package com.yahoo.athenz.zts.store;
+
+import java.util.*;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class RolePrefixTrie {
+    private final ReentrantReadWriteLock rwLock = new ReentrantReadWriteLock();
+    private final Lock readLock = rwLock.readLock();
+    private final Lock writeLock = rwLock.writeLock();
+
+    private final TrieNode root = new TrieNode();
+
+    private class TrieNode {
+        private HashMap<String, TrieNode> children = new HashMap<>();
+        private Set<String> roles = new HashSet<>();
+    }
+
+    public void insert(String prefix, String role) {
+        try {
+            writeLock.lock();
+            if (prefix.endsWith(".*")) {
+                prefix = prefix.substring(0, prefix.length() - 2);
+            }
+            TrieNode current = root;
+            final String[] prefixSections = prefix.split("\\.");
+            for (String domain: prefixSections) {
+                current = current.children.computeIfAbsent(domain, c -> new TrieNode());
+            }
+            // Add roles in the last node
+            current.roles.add(role);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    public void delete(String prefix, String role) {
+        try {
+            writeLock.lock();
+            if (prefix.endsWith(".*")) {
+                prefix = prefix.substring(0, prefix.length() - 2);
+            }
+            String[] prefixSections = prefix.split("\\.");
+            delete(root, prefixSections, role, 0);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    private boolean delete(TrieNode current, String[] prefixSections, String role, int index) {
+        if (index == prefixSections.length) {
+            current.roles.remove(role);
+            return current.roles.isEmpty() && current.children.isEmpty();
+        }
+        String section = prefixSections[index];
+        TrieNode node = current.children.get(section);
+        if (node == null) {
+            return false;
+        }
+        boolean shouldDeleteCurrentNode = delete(node, prefixSections, role, index + 1);
+
+        if (shouldDeleteCurrentNode) {
+            current.children.remove(section);
+            return current.roles.isEmpty() && current.children.isEmpty();
+        }
+        return false;
+    }
+
+    public Set<String> findRoles(String principal) {
+        try {
+            readLock.lock();
+            Set<String> roles = new HashSet<>();
+            TrieNode current = root;
+            final String[] principalSections = principal.split("\\.");
+
+            for (int i = 0; i < principalSections.length; i++) {
+                if (isLastSection(principalSections, i)) {
```

Comment:
why not just changing the for loop condition to `principalSections.length - 1` ?